### ISSUE TITLE
Android: KeyboardAvoidingView white-space on bottom of screen when android:windowTranslucentStatus is set to true

### ISSUE THREAD

**Issue by hrastnik** [facebook/react-native#27526]
KeyboardAvoidingView leaves space on bottom of screen when the keyboard is closed.
This bug manifests when the `android:windowTranslucentStatus` is set to `true`. (Note that this is the case for project initialized with Expo)

![kav-bug](https://user-images.githubusercontent.com/3407234/70899677-62037b80-1ff7-11ea-8ff8-567285704d27.png)

React Native version:
```
System:
    OS: macOS 10.15.2
    CPU: (4) x64 Intel(R) Core(TM) i5-4308U CPU @ 2.80GHz
    Memory: 68.76 MB / 8.00 GB
    Shell: 5.7.1 - /bin/zsh
  Binaries:
    Node: 12.4.0 - ~/.nvm/versions/node/v12.4.0/bin/node
    Yarn: 1.19.1 - ~/.nvm/versions/node/v12.4.0/bin/yarn
    npm: 6.13.1 - ~/.nvm/versions/node/v12.4.0/bin/npm
    Watchman: 4.9.0 - /usr/local/bin/watchman
  SDKs:
    iOS SDK:
      Platforms: iOS 13.2, DriverKit 19.0, macOS 10.15, tvOS 13.2, watchOS 6.1
    Android SDK:
      API Levels: 23, 26, 27, 28, 29
      Build Tools: 27.0.3, 28.0.3, 29.0.0
      System Images: android-28 | Intel x86 Atom_64, android-28 | Google APIs Intel x86 Atom, android-29 | Google APIs Intel x86 Atom
  IDEs:
    Android Studio: 3.5 AI-191.8026.42.35.5791312
    Xcode: 11.2.1/11B500 - /usr/bin/xcodebuild
  npmPackages:
    react: 16.9.0 => 16.9.0 
    react-native: 0.61.5 => 0.61.5 
  npmGlobalPackages:
    react-native-cli: 2.0.1
```
## Steps To Reproduce

1. Initialize a new RN project
2. Add `<item name="android:windowTranslucentStatus">true</item>` to `android/app/src/main/res/values/styles.xml`
3. Copy/paste the following code to `App.js`

```javascript
import React from 'react';
import {TextInput, View, KeyboardAvoidingView} from 'react-native';

export default () => (
  <KeyboardAvoidingView behavior="padding" style={{flex: 1}}>
    <View
      style={{flex: 1, justifyContent: 'center', backgroundColor: 'yellow'}}>
      <TextInput placeholder="Tap here to open keyboard" />
    </View>
  </KeyboardAvoidingView>
);
```
4. Run the app using `react-native run-android`
5. Open and close the keyboard
6. Observe the white-space on the bottom of the screen

Describe what you expected to happen:
It's expected for the yellow View to take up the whole space.

Snack, code example, screenshot, or link to a repository:
https://snack.expo.io/@hrastnik/keyboardavoidingview-bug

**Comment by patrickreck** [comment]
Have you found a solution to this?

**Comment by hrastnik** [comment]
Not really. I set the behaviour to height as it seems to work better sometimes, but it's not a solution.

I'm just waiting if a fix comes soon, otherwise I'll try rolling my own KeyboardAvoidingView.

**Comment by MitchellSlavik** [comment]
After some digging, this looks like it was caused by the keyboardDidHide event getting actual values on Android #24947. KeyboardAvoidingView is assuming that event will come back with no value and it will set it's bottom state value to 0. The values in the keyboardDidHide event are those from [View.getWindowVisibleDisplayFrame()](https://developer.android.com/reference/android/view/View.html#getWindowVisibleDisplayFrame(android.graphics.Rect)), which is the visible frame of the device without the statusbar since anything under the status bar would not be visible to the user. This causes the KeyboardAvoidingView to think that there is a keyboard the height of the statusbar still on the screen and causes the white line at the bottom.

**Comment by Drzaln** [comment]
Have anyone found a solution for this issue?

**Comment by dhruv-toptal** [comment]
Facing same issue on RN v0.60 and <StatusBar backgroundColor={'transparent'} translucent={true}  />

**Comment by devstarman** [comment]
Hi, is there any chance that this issue will be addressed soon?
I've switched to EXPO 36.0 recently and started to experience a lot of buggy behavior with KeyboardAvoidingView since :(

**Comment by MitchellSlavik** [comment]
@Drzaln @dhruv-toptal @devstarman
Since KeyboardAvoidingView is a purely js implementation, you can work around this issue in the mean time by running your own KeyboardAvoidingView. Here is what I am currently using it is just a slightly modified version of the one in react-native.
https://gist.github.com/MitchellSlavik/0cf342af5711c36d69af5224dba1f7e5

**Comment by hrastnik** [comment]
I found that behavior="height" works fine and doesn’t leave the white space on the bottom. 
However iOS works weird with behavior="height", so the solution is to use:

behavior={Platform.select({ android: "height", ios: "padding" })}

**Comment by devstarman** [comment]
@hrastnik you are almost right but unfortunately, I found a bug with behavior="height" on Android and posted it here: https://github.com/facebook/react-native/issues/28004

**Comment by josmontes** [comment]
Besides the whitespace, does anybody know a workaround on bottom tabs jumping when the keyboard closes?

**Comment by devstarman** [comment]
@MitchellSlavik I actually have not noticed your suggestion and I'm pleased to report that it's working fabulously! Thank you very much! The question remains why nobody from react-native team checked this issue reported in multiple posts if the fix is available and only needs to be included into next official release as a fix.

**Comment by giregk** [comment]
> @Drzaln @dhruv-toptal @devstarman
> Since KeyboardAvoidingView is a purely js implementation, you can work around this issue in the mean time by running your own KeyboardAvoidingView. Here is what I am currently using it is just a slightly modified version of the one in react-native.
> https://gist.github.com/MitchellSlavik/0cf342af5711c36d69af5224dba1f7e5

@MitchellSlavik did you or can you create a PR ?

**Comment by giregk** [comment]
> Besides the whitespace, does anybody know a workaround on bottom tabs jumping when the keyboard closes?

if you are using @MitchellSlavik 's solution, I found that instead of using his _onKeyboardHide method, if you keep using _onKeyboardChange and replace , at the end of that method, `this.setState({bottom: height})` with `this.setState({bottom: Platform.OS === "ios" ? height : 0})`, there is no jump

**Comment by t-fritsch** [comment]
Not sure how this is related, but I noticed the KeyboardAvoidingView also has problems of height/padding restoration when changing orientation on android : rotating the phone from portrait to landscape and then back to portrait results in unwanted spacing at the bottom of the view.

Using @MitchellSlavik hack seems to resolve this issue too ! thanks a lot 🙏

**Comment by barnu5** [comment]
Also having this issue

**Comment by Octaner** [comment]
Same problem with space on bottom of screen after upgrade RN from 0.60.6 to 0.61.0.
Bug disappears only after removing SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN and SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION

**Comment by fabriziobertoglio1987** [comment]
While preparing the pull request for this issue I post the following:

https://github.com/facebook/react-native/pull/24947 normalized **`keyboardDidHide`** return value as following:

![image](https://user-images.githubusercontent.com/24992535/86755613-0dcaf200-c042-11ea-9a28-e33f68452874.png)

but the `height` returned in the below line does not include the `StatusBar`

![image](https://user-images.githubusercontent.com/24992535/86756554-c133e680-c042-11ea-9228-b2f042c36564.png)

https://github.com/facebook/react-native/blob/e23e9328aa164d0a70fe4f16042c982e7801d924/ReactAndroid/src/main/java/com/facebook/react/ReactRootView.java#L687

I changed the height to 900 in my case and the white bar does not show

<img src="https://user-images.githubusercontent.com/24992535/86756819-f6403900-c042-11ea-8228-91ab72f28040.gif" width="250" />

I'm confident that I can publish pull request to fix this issue.

@MitchellSlavik is probably right saying that [getWindowVisibleDisplayFrame](https://developer.android.com/reference/android/view/View.html#getWindowVisibleDisplayFrame(android.graphics.Rect)) does not include the `StatusBar` in the computation, this seems to be only applicable when using `android:windowTranslucentStatus`.. I'll will find a solution for this issue and publish the pull request within days.

Thanks. Best Regards.

**Comment by sanchitos** [comment]
I was working on a pure JS RN app so the approach of @MitchellSlavik was not suitable for me.
If you have the same 'problem' this package saved my life: 
[https://www.npmjs.com/package/react-native-keyboard-aware-scroll-view](https://www.npmjs.com/package/react-native-keyboard-aware-scroll-view)

**Comment by wmonecke** [comment]
@MitchellSlavik Your implementation is not consistent in Android. Works on API 28 but not 29.

**Comment by fabriziobertoglio1987** [comment]
This issue was fixed with release [0.63.3](https://github.com/react-native-community/releases/blob/master/CHANGELOG.md#fixed-1) with my pr https://github.com/facebook/react-native/pull/29292

I quote the message from @andres-torres-marroquin in https://github.com/facebook/react-native/issues/30770#issuecomment-771876540
>I wonder about the behavior parity between iOS and Android, since on Android you could get multiple calls from keyboard events, if happens that you have more than one ReactRootView, I wonder if there is a way to do it as it is done on iOS, where you get the event triggered just once. (therefore depending of the app, might be a performance hit on Android)

I'm proposing to write a pr to `react-native-navigation` as explained in my post https://github.com/facebook/react-native/issues/30770#issuecomment-771723158

I would be happy to receive your feedback as I'm evaluating several potential solutions, including writing a pr to reverse this change or find an alternative solution. I will update https://github.com/facebook/react-native/issues/30770 with any new idea, please join the conversation. Thanks a lot :peace_symbol: :pray: 

>I'll send pr to `react-native-navigation` to override `checkForKeyboardEvents` in `ReactView`, if those are the views used for this scenario. 

```java
public class ReactView extends ReactRootView implements IReactView, Renderable {
    @Override
    protected void checkForKeyboardEvents() {
        super.checkForKeyboardEvents();
        // copy below logic
    }
 }
```

https://github.com/wix/react-native-navigation/blob/15c7a7e799f3cff9ad83b782dccc92aa1fe8a523/lib/android/app/src/main/java/com/reactnativenavigation/react/ReactView.java#L23-L43

>`checkForKeyboardEvents` should look like this

https://github.com/facebook/react-native/blob/98309fa7fce9e7d72416c76a25dc0f759f46c3e8/ReactAndroid/src/main/java/com/facebook/react/ReactRootView.java#L667-L697

**Comment by MayoudP** [comment]
> This issue was fixed with release [0.63.3](https://github.com/react-native-community/releases/blob/master/CHANGELOG.md#fixed-1) with my pr #29292

Does this version ( 0.63.3) and fixe suggestion fixed the issues in Android 11 where Keyboard doesn't show up in KeyboardAvoidingView ? Or does it seems to you a completely different bug ? thx

**Comment by fabriziobertoglio1987** [comment]
@MayoudP open and issue with a Minimum Reproducible Example and I'll have a look into it. Thanks a lot :+1:
